### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by thisismy-github**
New options:
- Single-click expandable folders
- Disable showing modified settings in bold
- Custom search hint
- Hide "see more results" item in search
- Custom commands for every type of control
- Single-expand options for every command in Windows 7 style
- Open pinned folders to their true path
- Customize pinned folder's location*

Other changes:
- Settings now detect whether or not they're actually modified
- Bug fix for loading XML files while an edit box is open
- "Open Pinned" added to start button's context menu
- New setting type for picking and saving directories called TYPE_DIRECTORY

*Because this changes the way the shell extension for pinning items through the context menu works, users may need to reinstall/repair their installations to actually update their context menus and start pinning to custom locations. I'm not sure if this will cause any issues when officially updating.

This is my first serious pull request, and my first serious C++ project. I tried my best to stay as consistent to the original code and formatting as possible. I wanted to do more, but finally stopped myself at 10 commits.

### CODE REVIEW

**Comment by bonzibudd** [comment]
@thisismy-github 

These seem like a good amount of useful changes, thank you for taking the time to do this. Everything seems to work as I expected, except for one thing: The "Single-click to open folder shortcuts" option does not seem to be able to differentiate between jump list items and normal sub-menu lists. This causes the app to open directly when clicking a jump list, making those items inaccessible. 

Other than that, everything seems to be working as expected, except for some remaining issues with the "Pinned" items location and switching between the pinned folder in the "Programs" list to the standard "Pinned" folder.

If you want to go further into those issues, I would take a look at #532 and #522, which outline some of those problems, mainly when the `Use Start Menu folder` option is enabled (default in classic styles).

Again, thank you for taking your time to do this (Although I am not a maintainer, I have been using OSM for a while). I will try to post some more information if I find any problems with the custom Pinned folder, or anything else that may cause issues.

**Comment by thisismy-github** [comment]
@bonzibudd Good catch, thank you! Fixed. Also just to be clear for the issues you mentioned about the pinned location, are those specifically problems my changes caused/worsened or are they just related things you wanted me to look into? Because to be honest, I have no idea what causes those problems, and they seem like they may be deceptively complicated or possibly very deep-seated. I'll probably look into those eventually though (assuming they're not my fault and this pull request goes smoothly), after I finish a few other issues and ideas I have lined up.

Also, I wanted to thank you (a second time). [The change I did for Fluent-Metro](https://github.com/bonzibudd/Fluent-Metro/pull/162) two weeks ago was my first ever pull request, and you merging it (even after how hard I bungled it initially) is what gave me the confidence to try and tackle something as big as this, so really: Thank you. But also... you merging my first pull request indirectly resulted in my second pull request adding global search hint customization, thus making the first one obsolete. A cruel twist of irony, but a worthwhile sacrifice. So thank you (a fourth time).

**Comment by blackcrack** [comment]
hmm.. not need to  "seals around in the mud" we we are be in the international "land" ;) there is one grateful thank you enough ;)
弓 

>If you want to go further into those issues, I would take a look at #532 and #522, which outline some of those problems, mainly when the Use Start Menu folder option is enabled (default in classic styles).  

a bugfix too of this suggestions ?
best

**Review by blackcrack — APPROVED** [review]
huhh, who have set up the "&&" .. we are not in dos batch..
thy for support our community :))

**Review by blackcrack — APPROVED** [review]

**Comment by bonzibudd** [comment]
@thisismy-github 

From what I can tell so far, those specific issues haven't been worsened. The main reason I brought it up was just because changing those settings may have some unintended consequences regarding issues like that, especially if someone didn't know of the strange behavior beforehand. However, since we haven't found any similar problems at this point, I definitely wouldn't worry about it, unless you had looked into it already.

One more thing: For the `Single-click to open folder shortcuts` option, I would suggest also excluding the Metro "Apps" folder, which is shown on the classic menu styles. Because of the way that folder works, you can't actually click to open it in a window, which makes the list unusable unless you just hover over it.
image


Again, thank you for the contributions!

**Comment by thisismy-github** [comment]
@bonzibudd Fixed. I noticed that you said the apps folder can't be opened in a window at _all_, so on top of fixing the single-click issue, I also added a change to prevent it from opening via double clicks either so you can never accidentally "open" it. It still expands when you click it. I hope I didn't misunderstand you and that this change wasn't too obtrusive.

**Review by blackcrack — APPROVED** [review]

**Comment by blackcrack** [comment]
@ge0rdi common, check it :)

**Comment by bonzibudd** [comment]
Sounds good! I'm not at my computer right now, so I can't test the changes immediately, but I'll post an update about the double-clicking behavior when I am able to.

**Comment by bonzibudd** [comment]
@thisismy-github 

The clicking behavior with the Apps item seems much better now. I personally think it makes the most sense as it is now, as it wouldn't be intuitive to be able to click the item and wait for it to not do anything.

**Comment by Gaurav-Original-ClassicShellTester** [comment]
Hello. I just noticed your contributions and finally had time to test them. Thank you so much for enhancing this great project's features. Here are my comments:

● Single-click expandable folders
- Similar to the Apps folder, if I add the Manage Wireless Networks folder (shell:::{1FA9085F-25A2-489B-85D4-86326EEDCD87} as a LINK item, that too causes a problem with single click. Because since Windows 8, Microsoft broke this shell folder to open in Explorer so it just closes the menu instead of expanding it. 

Another example: if I add recent_programs as a submenu-item COMMAND in the Classic styles (instead of in the main menu), it also just closes the menu.

3rd example: for items in the Classic styles like "Search", when single click is turned on, and this item also isn't supposed to open anything, it should immediately expand the menu upon click (instead of waiting for Menu delay) because that's one of the Windows features - click signifies open any submenu immediately, without delay even if there is a defined delay value.

Is there a way to fix it in a generic way for things that do not open in Explorer, instead of excluding them individually? :)

● Disable showing modified settings in bold
- Note that there was a reason why a modified setting turns bold but unmodifying it does not revert to non-bold. It is a user-feature for more control over exporting the XML. You see, when you export an XML without touching any setting, only the modified settings are exported. When you modify a setting, it means it will be exported in the XML. If you touch it and revert it to the default value and it remains bold, it is still considered modified and implied that it must still be exported in the XML but with the modified value. Only if you reset it to default, it will not be exported again in the XML. So your change seems to take away this feature? There will be no way to export non-default values to the XML.

● Custom search hint
- Nice :)

● Hide "See more results" item in search
- I requested this multiple times to Ivo but he refused because I was already requesting way too many features and by the time I had made this request, there were already too many settings. 😆😆😆 Initially Ivo was against providing many settings and I literally argued with him that Microsoft does the same thing - we need a power user tool to fix Windows GUI limitations and convinced him completely against his initial plan to have a very dumbed down menu with hardly any settings to customize.  So thank you thank you thank you for adding it! :) There is no such thing as too many settings, haha.

Btw the implementation of "See more results" in Classic Shell is incomplete. Unlike the original Start menu in Windows 7, it does not support Advanced Query Syntax. So if you type for example:

kind:images ext:png
size:gigantic
kind:video
date:pastweek

you get results in the menu, but not in Explorer when you click "See more results". This is because Ivo is using the search API instead of a URL. Instead what I do in my custom menu setup on my PC is I use a search provider URL (http://www.classicshell.net/forum/viewtopic.php?f=4&t=1588#SearchProviders). That URL is:

search-ms:&crumb=location:::{9343812E-1C37-4A49-A12E-4B2D810D956B}&crumb=%2

When this is used, it shows the same results in Explorer which the menu shows (this GUID searches indexed locations.search-ms, you can also use it to search specific locations e.g. search-ms:&crumb=location:C:\&crumb=%2 - a search provider added as "Search C: drive").

Maybe you can switch to using this URL instead of the search API so the Explorer results are same as the menu.

● Custom commands for every type of control
- BRILLIANT! Thank you! Looking forward to your Shift+middle click and your semi-big change. :) Another discarded idea to Ivo I gave was middle click inside the menu (since it is unused) on any item to "Pin" and middle click on anything in the Pinned list to unpin. Another big improvement would be allowing adding Custom commands for the "Shutdown" in Windows 7 style.

● Single-expand options for every command in Windows 7 style
Awesome! This is great for any sub-folder item that takes a while to populate (which should be opened in Explorer anyway). Btw, it doesn't show up for Devices and Printers. Also I feel it should not show for any external command we use in the COMMAND field of the Edit Menu Item dialog.

● Open pinned folders to their true path
Terrific! I remember many users asking for this enhancement in the Classic Shell forum and Sourceforge before that. :))

● Customize pinned folder's location
Amazing! Seems to work great as long as the folder has permissions. Also, none of the issues with jumplists like the default Start menu folder has, which Bonzibudd found :)

● "Open Pinned" added to start button's context menu
Great change.

● Bug fix for loading XML files while an edit box is open
Thank you

**Comment by ge0rdi** [comment]
> If you touch it and revert it to the default value and it remains bold, it is still considered modified and implied that it must still be exported in the XML but with the modified value. Only if you reset it to default, it will not be exported again in the XML. So your change seems to take away this feature? There will be no way to export non-default values to the XML.

I tend to agree with this.

Basically by changing setting to value X you are telling that you want this setting to be at value X no matter of what.
Sure, X may be current default value for the setting, but it may happen that the default value will change in future update of `Open-Shell`. And then your choice won't be respected (as the setting will no longer be at value X).

If user wants to set setting to default value (whatever that is and will be in the future) there is `Set default` option in context menu that will achieve that.

I think this particular change is not desired.

**Comment by ge0rdi** [comment]
@thisismy-github 
Wondering what reasoning is behind `Added option to never highlight modified settings` change.

Why would this be ever needed?
What is the problem with showing changed settings in bold?

File: Src/StartMenu/StartMenuDLL/MenuContainer.cpp
Comment (ge0rdi) [inline Src/StartMenu/StartMenuDLL/MenuContainer.cpp:448]: It would be better to have local variable that will hold actual string.
And then call `DrawText` at one place (with proper string).
So that we don't need to duplicate all the parameters (and risk they will get out of sync in the future).

**Comment by ge0rdi** [comment]
@thisismy-github 
I'd like to ask you to use `git commit --fixup <commit-id>` for fixup commits (such as `Various single-click option fixes` one) use.

This way it will be clear what _base_ commit given fixup is related to. And also those fixups can be easily applied (amended) using `git rebase -i --autosquash ...` before PR is merged.

The thing is that in the end we want to keep only _real_ (relevant) commits and not some remnants of development (fixups).
Thank you for understanding.

**Review by ge0rdi — CHANGES_REQUESTED** [review]

**Comment by ge0rdi** [comment]
There is some issue with disabled `See more results` option in Windows 7 style:
image
Looks like the last line is not shifted properly.

**Comment by Gaurav-Original-ClassicShellTester** [comment]
> There is some issue with disabled `See more results` option in Windows 7 style:
> Looks like the last line is not shifted properly.

Yes when "Search files" is unchecked, this is occurring in Windows 7 style.

**Comment by ge0rdi** [comment]
@thisismy-github 
> I tried my best to stay as consistent to the original code and formatting as possible.

Really appreciate that.
Many people don't understand that consistency is very important.

> I wanted to do more, but finally stopped myself at 10 commits.

I'd personally prefer more PRs where each deals with one feature (area of interest).
It is easier to discuss single thing (or couple tightly related changes) than multiple unrelated changes.

So don't worry and submit more PRs in the future.

**Comment by ge0rdi** [comment]
> Yes when "Search files" is unchecked, this is occurring in Windows 7 style.

For me it happens even if `Search files` is enabled (e.g. default settings).

**Comment by Gaurav-Original-ClassicShellTester** [comment]
Can "Single-expand options" (Display as a link of links) be expanded to Classic styles as well - Inside the "Edit Menu Item" dialog?

**Comment by muhachris297** [comment]
how does one go about downloading versions such as yours that were not merged with the official html files?

**Comment by VarunAgw** [comment]
@muhachris297 https://github.com/Open-Shell/Open-Shell-Menu/issues/691#issuecomment-952755008

**Comment by e33et** [comment]
I really like Open-Shell Menu cause have been user since Classic-Shell Menu, but "EAC Easy Anti-cheat" seems doesn't like it, cause it got broken after I launched Star Citizen or Halo, that use EAC as anti-cheat!

Hope that u could fix it for e.g whitlist or flag it, as safe 3rd party apps!

Otherwise it's working flawless!

Regards
/e33et

**Review by Ibuprophen — APPROVED** [review]
This does look like a logical choice at this point in time @ge0rdi. 👍

**Review by Ibuprophen — APPROVED** [review]
This does have potential @ge0rdi... 🥇

**Comment by bonzibudd** [comment]
@thisismy-github Any updates on this? I would like to see this eventually pushed, but I understand if you are busy or working on other things.

**Comment by blackcrack** [comment]
@ge0rdi what's goes on, Covit-19 corpse ?

best

@Ibuprophen ..hmmm ?
do you know anything about him ? he's in the net or "past" ? could you mail him ?

**Comment by coddec** [comment]
> @ge0rdi what's goes on, Covit-19 corpse ?
> 
> best
> 
> @Ibuprophen ..hmmm ? do you know anything about him ? he's in the net or "past" ? could you mail him ?

Hi @blackcrack 
Please be respectful and polite in the community while communicating, no one should be threatened or even made uncomfortable intentionally.
covid is not something fun, it's even worse to use it as a joke against someone.
To answer your question, we are all volunteers, we pretty much have no communication or connection at all out side of this project (Not as far as I know of).
Everyone has their own life and has to make living as well plus other matters to deal with. Maybe that's what @ge0rdi up to.
Thanks for your input again, but such joke and language can be considered unrespectful, impolite or offensive.
Hopefully you can change the way asking.
If continuously making the community feeling offensive, impolite or unrespestful, necessary action will have to be taken. e.g. report to github, block etc.
Wish you all the best and see you around hopefully with polite comments, let's make it a safe and joyful community.
Thanks

**Comment by ge0rdi** [comment]
Thank you, @coddec .

I'm no longer participating in the project because I'm simply not interested in it that much anymore.
I found that I can really get along with standard Windows start menu, so I'd rather keep my energy for other projects.

Though I have no problem to review PRs like this from time to time, or provide some advice (if I can help somehow).

Regarding current PR, I have requested some changes and it is now up to @thisismy-github to implement them or discuss them.

**Comment by coddec** [comment]
> Thank you, @coddec .
> 
> I'm no longer participating in the project because I'm simply not interested in it that much anymore. I found that I can really get along with standard Windows start menu, so I'd rather keep my energy for other projects.
> 
> Though I have no problem to review PRs like this from time to time, or provide some advice (if I can help somehow).
> 
> Regarding current PR, I have requested some changes and it is now up to @thisismy-github to implement them or discuss them.

Thanks for your reply @ge0rdi 
Totally understandable.

That will be great, as you may still come back from time to time, hope we will see you around the corner :)
Appreciate letting us know.
Wish you the best!

**Comment by Ibuprophen** [comment]
Not to be repetitive @XenHat... LOL! 

Just a bit of feedback as to this push have everything it needs to work via the Team. 👍 

~Ibuprophen

**Review by blackcrack — APPROVED** [review]

**Review by Ibuprophen — APPROVED** [review]

**Comment by ge0rdi** [comment]
@thisismy-github 

One of your changes broke menu sorting. See #1111 for details.
I have reverted some of those changes (`bFolderLink` related) in PR #1128.

Honestly, I don't get what was the point of those changes.
Feel free to get them back, if you think they are really needed. Just make sure it won't break the sorting 😄

**Comment by ge0rdi** [comment]
@thisismy-github 

I have reverted the `Auto-detect when setting is back to default value` change as it was wrong (see https://github.com/Open-Shell/Open-Shell-Menu/pull/829#issuecomment-907667829).
It is unfortunate that this PR got merged without sorting that out 😞

**Comment by wmjordan** [comment]
I appreciate @thisismy-github's commits and contributions. It is nice to have some new features.
Can the newly settings be added to the localization template dll file, so they can be localized to other languages?

**Pull request merged**
